### DISCUSSION TITLE
Maven toolchains support

### DISCUSSION THREAD

**Discussion by mkpaz** [wiverson/jtoolprovider-plugin#4] [General]
Some OpenJDK vendors like Liberica or Zulu provide OpenJDK builds with bundled OpenJFX. This is useful if someone wants to build its app for OS arch that OpenJFX doesn't support itself (e.g. for x32 Windows systems).

To create native images in a cross-platform fashion one have to specify corresponding jmods source. From what I've learned from [jpackage-template](https://github.com/wiverson/maven-jpackage-template/blob/main/pom.xml#L161) you extracted jmods from OpenJFX dist and keeping them with code in the same repo. I'm not sure, but it looks like this may result to error if someone will try to create native Linux image from Windows. In that case jlink will always pickup Java jmods from Windows version and OpenJFX jmods from target platform version. 

[Maven Toolchains](https://maven.apache.org/guides/mini/guide-using-toolchains.html) is merely a provider to locate Java home directory based on some filter criteria. So one can use it to build for different platforms and against different JDK versions from the single machine. [Here ](https://github.com/moditect/moditect/blob/master/maven-plugin/src/main/java/org/moditect/mavenplugin/image/CreateRuntimeImageMojo.java#L116) is an example how toolchains support is implemented in [Moditect](https://github.com/moditect/moditect#creating-modular-runtime-images).  It basically just searches jmods dir based on JDK path specified and [includes](https://github.com/moditect/moditect/blob/master/core/src/main/java/org/moditect/commands/CreateRuntimeImage.java#L152) it to module path. It would be really nice if JToolProvider supported similar option in addition to `modulePath`. Like, when `baseJdk` is specified its jmod dir is added to the specified in `modulePath` value.

In any case thank you very much for your work. JToolProvider may be small but also super flexible as it works like simple CLI wrapper.

**Comment by wiverson** [comment]
Right now, AFAIK jlink automatically includes the JDK path for modules that it's invoked from.

RE: JavaFX builds, you should check out this project: https://github.com/wiverson/maven-jpackage-template

That template uses Maven platform detection to automatically pick the correct set of JavaFX jmods. jpackage doesn't support builds across platforms - you have to use Linux to build Linux, Windows to build Windows, macOS to build macOS. This is going to get a bit more complicated going forward with JDK 17 as Linux ARM/macOS M1 builds make things more confusing. I assume in the near future it will be x64/Intel and ARM across Linux, macOS and Windows. Whee!  ;)

**Reply by wiverson** [reply]
There is a world of complexity that folks could go down, e.g. trying to run Maven on JDK A while targeting JDK B. TBH I'm not sure if that even makes sense.

Found it - https://docs.oracle.com/en/java/javase/16/docs/specs/man/jlink.html

"If this option is not specified, then the default module path is $JAVA_HOME/jmods. This directory contains the java.base module and the other standard and JDK modules. If this option is specified but the java.base module cannot be resolved from it, then the jlink command appends $JAVA_HOME/jmods to the module path."
